python/ray/data/dataset.py

```diff
@@ -747,16 +758,39 @@ def add_column(
                 ray (e.g., num_gpus=1 to request GPUs for the map tasks).
         """
 
-        def add_column(batch: "pandas.DataFrame") -> "pandas.DataFrame":
-            batch.loc[:, col] = fn(batch)
-            return batch
+        def add_column(
+            batch: DataBatch,
+        ) -> Union["pyarrow.Array", "pandas.Series", Dict[str, "np.ndarray"]]:
+            column = fn(batch)
+            if batch_format == "pandas":
+                batch.loc[:, col] = column
+                return batch
+            elif batch_format == "pyarrow":
+                # Historically, this method was written for pandas batch format.
+                # To resolve https://github.com/ray-project/ray/issues/48090,
+                # we also allow pyarrow batch format which is preferred but would be
+                # a breaking change to enforce.
+
+                # For pyarrow, the index of the column will be -1 if it is missing in
+                # which case we'll want to append it
+                column_idx = batch.schema.get_field_index(col)
+                if column_idx == -1:
+                    # Append the column to the table
+                    return batch.append_column(col, column)
+                else:
+                    # Create a new table with the updated column
+                    return batch.set_column(column_idx, col, column)
+            else:
+                # batch format is assumed to be numpy
```

Comment:
Let's not assume and instead add explicit conditional (for unsupported format throw an exception)

Reply:
While I'm fine with that, this collides with https://github.com/ray-project/ray/pull/48140#discussion_r1842667475

Reply:
I'll revert the change I made for Richard then, assuming that I should follow the recommendation of the Ray Data team here.

Reply:
Yeah, let's not assume the format -- UDF has to match the format and hence we need to be careful with an assumptions like that